src/ecmult_const_impl.h

```diff
@@ -12,220 +12,242 @@
+     *
+     * It is appealing to try to combine this with the GLV optimization; the idea that a scalar
+     * a can be written as s1 + lambda*s2, where lambda is a curve-specific constant such that
+     * lambda*A is easy to compute, and where s1 and s2 are small. In particular we have the
+     * secp256k1_scalar_split_lambda function which performs such a split with the resulting s1
+     * and s2 in range (-2^128, 2^128) mod n. This does work, but is uninteresting:
+     *
+     *   To compute q*A:
+     *   - Let s1, s2 = split_lambda(q)
+     *   - Let R1 = C_256((s1 + 2^256 - 1) / 2, A)
+     *   - Let R2 = C_256((s2 + 2^256 - 1) / 2, lambda*A)
+     *   - Return R1 + R2
+     *
+     * The issue is that while q1 and q2 are small-range numbers, (q1 + 2^256 - 1) / 2 mod n
+     * and (q2 + 2^256 - 1) / 2 mod n are not, undoing the benefit of the splitting.
+     *
+     * The make it work, we apply the corrections to the input scalar, including the division by
+     * two, before calling split_lambda. Let's introduce an offset k, and solve for it. We also
+     * add 2^128-1 to s1 and s2 after splitting to make sure we end up with positive scalars.
+     *
+     *   To compute q*A:
+     *   - Let s1, s2 = split_lambda((q + k) / 2 mod n)
+     *   - Let v1 = s1 + 2^128 - 1
+     *   - Let v2 = s2 + 2^128 - 1
+     *   - Let R1 = C_l(v1, A)
+     *   - Let R2 = C_l(v2, lambda*A)
+     *   - Return R1 + R2
+     *
+     * l will thus need to be at least 129, but we may overshoot by a few bits (see
+     * further), so keep it as a variable.
+     *
+     * To solve for k, we reason:
+     *     q*A = R1 + R2
+     * <=> q*A = C_l(s1 + 2^128 - 1, A) + C_l(s2 + 2^128 - 1, lambda*A)
+     * <=> q*A = (2*(s1 + 2^128 - 1) + 1 - 2^l) * A + (2*(s2 + 2^128 - 1) + 1 - 2^l) * lambda*A
+     * <=> q*A = (2*(s1 + s2*lambda) + (2^129 - 1 - 2^l)*(1 + lambda)) * A
+     * <=> q   = 2*(s1 + s2*lambda) + (2^129 - 1 - 2^l)*(1 + lambda) mod n
+     * <=> q   = 2*(q + k) / 2 + (2^129 - 1 - 2^l)*(1 + lambda) mod n
+     * <=> q   = q + k + (2^129 - 1 - 2^l)*(1 + lambda) mod n
+     * <=> k   = (2^l + 1 - 2^129)*(1 + lambda) mod n
+     *
+     * We will process the computation of C_l and C_l in groups of ECMULT_CONST_GROUP_SIZE, so we
+     * set l to the smallest multiple of ECMULT_CONST_GROUP_SIZE that is not less than 129; this
+     * equals ECMULT_CONST_BITS.
+     *
+     * All together, that means k = (2^ECMULT_CONST_BITS + 1 - 2^129) * (1 + lambda) (mod order).
+     */
 
-static void secp256k1_ecmult_const(secp256k1_gej *r, const secp256k1_ge *a, const secp256k1_scalar *scalar, int size) {
-    secp256k1_ge pre_a[ECMULT_TABLE_SIZE(WINDOW_A)];
-    secp256k1_ge tmpa;
+    /* The offset to add to s1 and s2 to make them positive. Equal to 2^128 - 1. */
+    static const secp256k1_scalar S_OFFSET = SECP256K1_SCALAR_CONST(0, 0, 0, 0, 0xfffffffful, 0xfffffffful, 0xfffffffful, 0xfffffffful);
```

Comment:
It seems 2^128 would work just as well for the offset. Is it easier to just add 1 to bit 128 than keep this offset around?

Reply:
Also ~~positive~~ non-negative here.